packages/contracts/contracts/maci/MaxCapVoiceCreditProxy.sol

```diff
@@ -0,0 +1,218 @@
+    totalAmount += amount;
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function withdrawExtra(
+    address[] calldata receivers,
+    uint256[] calldata amounts,
+    IPoll poll
+  ) public override onlyOwner whenNotPaused afterCooldown(poll) {
+    uint256 amountLength = amounts.length;
+    uint256 totalFunds = totalAmount;
+    uint256 sum = 0;
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      sum += amount;
+
+      if (sum > totalFunds) {
+        revert InvalidWithdrawal();
+      }
+
+      totalAmount -= amount;
+
+      unchecked {
+        index++;
+      }
+    }
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      address receiver = receivers[index];
+
+      if (amount > 0) {
+        token.safeTransfer(receiver, amount);
+      }
+
+      unchecked {
+        index++;
+      }
+    }
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function claim(
+    uint256 index,
+    uint256 voiceCreditsPerOption,
+    uint256 tallyResult,
+    uint256 totalVotesSquares,
+    uint256 totalSpent,
+    IPoll poll
+  ) public override whenNotPaused afterVotingDeadline(poll) {
+    uint256 amount = getAllocatedAmount(voiceCreditsPerOption, totalVotesSquares, totalSpent, tallyResult);
+    totalAmount -= amount;
+
+    IRecipientRegistry registry = poll.getRegistry();
+    IRecipientRegistry.Recipient memory recipient = registry.getRecipient(index);
+
+    token.safeTransfer(recipient.recipient, amount);
+  }
+
+  /// @notice Get allocated token amount (without verification).
+  /// @param voiceCreditsPerOption The voice credit options received for recipient
+  /// @param totalVotesSquares The sum of tally results squares
+  /// @param totalSpent The total amount of voice credits spent
+  /// @param tallyResult The tally result for recipient
+  function getAllocatedAmount(
+    uint256 voiceCreditsPerOption,
+    uint256 totalVotesSquares,
+    uint256 totalSpent,
+    uint256 tallyResult
+  ) internal view returns (uint256) {
+    uint256 alpha = calculateAlpha(totalVotesSquares, totalSpent);
+    uint256 quadratic = alpha * voiceCreditFactor * tallyResult * tallyResult;
+    uint256 totalSpentCredits = voiceCreditFactor * voiceCreditsPerOption;
+    uint256 linearPrecision = ALPHA_PRECISION * totalSpentCredits;
+    uint256 linearAlpha = alpha * totalSpentCredits;
+
+    return ((quadratic + linearPrecision) - linearAlpha) / ALPHA_PRECISION;
+  }
+
+  /// @notice Calculate the alpha for the capital constrained quadratic formula
+  /// @dev page 17 of https://arxiv.org/pdf/1809.06421.pdf
+  /// @param totalVotesSquares The sum of tally results squares
+  /// @param totalSpent The total amount of voice credits spent
+  function calculateAlpha(uint256 totalVotesSquares, uint256 totalSpent) internal view returns (uint256) {
```

Comment:
I do not quite understand this function

Reply:
Here is information about this formula: https://arxiv.org/pdf/1809.06421.pdf (page 17)